sakuya/wordle.py

```diff
@@ -78,7 +78,7 @@ class GuildState:
     last_guess_at: datetime = datetime.utcfromtimestamp(0)
     guesses: list[str] = None
     guessers: set[Member] = None
-
+    in_progress: bool = False
```

Comment:
I'm not a huge fan of adding state that can be (mostly) inferred by other state. Having to keep these variables in sync is additional overhead and might result in some unexpected bugs if things go wrong... but note that this variable can't be **entirely** inferred from other state (see "state change" below).

There's already logic to check whether a game has finished or not - effectively "not in progress":
https://github.com/Trif4/Sakuya/blob/5ac749d0226acef00748948ee30694a71b264ae1/sakuya/wordle.py#L133
so perhaps we could refactor that out into a `finished()` method on `GuildState`? Beware that `self.word` and `self.guesses` could be `None` at startup, despite the type annotations, so you'll need to check for that.

Note that a state change *is* necessary after sending the "times up!" message (or else Sakuya will keep on sending it). There's at least two ways of doing so, assuming `finished()` is refactored out:

- Introduce a similar boolean property, `times_up_message_was_sent` (any suggestions for a better name?), that records what it says. This removes the need to keep track of the current state of the game, but still introduces a new property.
- Make the "times up!" logic modify the state so it becomes finished (for example, append the `word` to `guesses`). This removes the need to introduce new state/properties to `GuildState`, but results in much more confusing code (changing a random variable for the sole purpose of making `finished()` return `True`).

What do y'all think? I'm leaning more towards the first one for readability and maintainability.

Reply:
During my first attempts, I wanted to deduce whether the message is needed or not from the existing state (particularly `word` being None).

It didn't work out, just as you predicted. Later parts of the function depend on `word` not being None and there's also no clean fit for keeping track of whether the message was sent or not. Then I went with the new state field to reduce the scope of my changes.

---

I think introducing the new state as `times_up_message_was_sent` is a viable change even without refactoring anything else in the function/GuildState.